### PR TITLE
[SYCL] Augment known_identity for std::complex and std::plus.

### PR DESCRIPTION
**Pull request by uditagarwal97**
This PR proposes to augment sycl::has_known_identity to return true for std::complex and std::plus operator. This will have two benefits:

1. It enables support for complex numbers in sycl::reduction without the user having to explicitly pass identity.
2. sycl::known_identity can now be used to simplify the implementation of group algorithms (See PR #5394).

Also, this PR addresses the Github issue #5477.
Test Case PR: https://github.com/intel/llvm-test-suite/pull/1609

### CODE REVIEW

File: sycl/include/sycl/known_identity.hpp [hunk 1]
```diff
@@ -63,14 +63,24 @@ using IsLogicalOR =
     bool_constant<std::is_same<BinaryOperation, sycl::logical_or<T>>::value ||
                   std::is_same<BinaryOperation, sycl::logical_or<void>>::value>;
 
+template <typename T>
+using isComplex =
+    bool_constant<std::is_same<T, std::complex<short>>::value ||
+                  std::is_same<T, std::complex<int>>::value ||
```
Comment (Pennycook) [inline sycl/include/sycl/known_identity.hpp:69]: Why do we have these two instantiations of `std::complex<>` for integral `T`?

I've only ever seen `std::complex<float>`, `std::complex<double>` and `std::complex<long double>`.  Both  [`sycl_ext_oneapi_complex`](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc) and [`sycl_ext_oneapi_complex_algorithms`](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc) only support `float`, `double` and `sycl::half`.

Comment (uditagarwal97) [reply sycl/include/sycl/known_identity.hpp]: Thanks for bringing this up. I wasn't aware that for std::complex, we only support float and double types. I've accordingly made the changes.

**Comment by aelovikov-intel** [comment]
Do we need to update any extensions with new wording?

**Comment by uditagarwal97** [comment]
> Do we need to update any extensions with new wording?

I'm not sure but I think update to [sycl_ext_oneapi_complex_algorithms.asciidoc](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc) is not required as it already introduces std::complex<float> and std::complex<double> to reduce and reduce_over_group algorithms. 
 @Pennycook What's your opinion?

**Comment by Pennycook** [comment]
> I'm not sure but I think update to [sycl_ext_oneapi_complex_algorithms.asciidoc](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc) is not required as it already introduces std::complex and std::complex to reduce and reduce_over_group algorithms. @Pennycook What's your opinion?

SYCL 2020 says:
> If an implementation can identify an identity value for a given combination of accumulator type and function object type, the value is defined as a member of the known_identity trait class. Whether this member value exists can be tested using the has_known_identity trait class.

Since the intent of this trait is to let people test which types an implementation can reason about, it seems redundant to also put that information into an extension specification.  But I might just be feeling lazy, so I'd like @gmlueck's opinion on this.

**Comment by gmlueck** [comment]
> Since the intent of this trait is to let people test which types an implementation can reason about, it seems redundant to also put that information into an extension specification. But I might just be feeling lazy, so I'd like @gmlueck's opinion on this.

I think it's worthwhile to add this to the extension specification.  The core SYCL spec has [Table 123 "Known identities"](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.identities), which lists the operators that have known identifies for certain types.  It would make sense for our extension spec to list extended cases that have known identities.

BTW, [sycl_ext_oneapi_complex_algorithms](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc) is listed as "proposed".  Is this implemented now?  If so, it should be moved to either "experimental" or "supported" as appropriate (and it should be updated to note its current status).

**Comment by uditagarwal97** [comment]
> > Since the intent of this trait is to let people test which types an implementation can reason about, it seems redundant to also put that information into an extension specification. But I might just be feeling lazy, so I'd like @gmlueck's opinion on this.
> 
> I think it's worthwhile to add this to the extension specification. The core SYCL spec has [Table 123 "Known identities"](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.identities), which lists the operators that have known identifies for certain types. It would make sense for our extension spec to list extended cases that have known identities.
> 
Sure, thanks for the feedback. I've updated the extension specification as well. @Pennycook please review the changes.

> BTW, [sycl_ext_oneapi_complex_algorithms](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc) is listed as "proposed". Is this implemented now? If so, it should be moved to either "experimental" or "supported" as appropriate (and it should be updated to note its current status).

I don't know if sycl_ext_oneapi_complex_algorithms is completely implemented.

**Comment by gmlueck** [comment]
> I don't know if sycl_ext_oneapi_complex_algorithms is completely implemented.

Do either of you know?  I'm wondering if we just forgot to move the extension specification.

**Comment by Pennycook** [comment]
> Do either of you know? I'm wondering if we just forgot to move the extension specification.

sycl_ext_oneapi_complex_algorithms says that it depends on two other extensions.  The first is [sycl_ext_oneapi_complex](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc), which is still showing as "proposed".  The second is sycl_ext_oneapi_marray, which hasn't actually been merged yet: https://github.com/intel/llvm/pull/6550

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc [hunk 2]
```diff
@@ -125,6 +125,10 @@ NOTE: `sycl::bit_and`, `sycl::bit_or`, `sycl::bit_xor`, `sycl::logical_and`,
 because their behaviors are defined in terms of operators that `std::complex`
 does not implement.
 
+=== Known Identities
+
+Along with the known identities listed in https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.identities[Table 123], `sycl::known_identity` and `sycl::has_known_identity` also support `std::complex` with the `float` and `double` floating-point types.
```
[the remark this answers is not in the document]
Comment (uditagarwal97) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc]: Thanks. Fixed!

**Review by gmlueck — APPROVED** [review]
spec changes LGTM

**Review by aelovikov-intel — APPROVED** [review]
SYCL RT changes LGTM.

**Comment by uditagarwal97** [comment]
@intel/llvm-gatekeepers The PR is ready!

**Comment by steffenlarsen** [comment]
CUDA Failed Tests (1):
    SYCL :: SubGroup/shuffle_fp64.cpp
Unrelated and reported in https://github.com/intel/llvm/issues/8516.

**Pull request merged**